### PR TITLE
Fix resurrection logic to trigger early enough

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137267

Fixes https://github.com/pytorch/pytorch/issues/136358

The bug here is that the Tensor object is actually 2 classes: `Tensor` from `_tensor.py` and `TensorBase` from c++.

Before this PR, they have the following gc methods:
Tensor:
 - tp_clear subtype_clear
 - tp_traverse THPVariable_subclass_traverse
 - tp_dealloc THPVariable_subclass_dealloc

TensorBase:
- tp_clear THPVariable_clear
- tp_traverse THPFunction_traverse (fake function that is just an error)
- tp_dealloc object_dealloc

The problem is that when clear is called on the Tensor, subtype_clear is going to clear the things owned by the "Tensor" type, in particular, its `__dict__` attribute, before delegating to the TensorBase clear where we detect that resurrection needs to happen and skip it.

### CODE REVIEW

File: torch/csrc/autograd/python_variable.cpp [hunk 1]
```diff
@@ -2277,9 +2283,17 @@ int THPVariableMetaType_init(PyObject* cls, PyObject* args, PyObject* kwargs) {
...
+  // resurrection checks to properly happen. In particular, an older version
+  // was not overriding tp_clear here. This lead to the default subtype_clear
+  // running on the Tensor object (as only TensorBase tp_clear was custom),
+  // clearing the __dict__ field, before the TensorBase custom clear was called
+  // and would properly detect the resurrect.
+  // See https://github.com/pytorch/pytorch/issues/136358 for the exact behavior
   ((PyTypeObject*)cls)->tp_dealloc = (destructor)THPVariable_subclass_dealloc;
   ((PyTypeObject*)cls)->tp_traverse =
       (traverseproc)THPVariable_subclass_traverse;
+  ((PyTypeObject*)cls)->tp_clear = (inquiry)THPVariable_clear;
```
Comment (kshitij12345) [inline torch/csrc/autograd/python_variable.cpp:2296]: I was wondering if this should be renamed to `THPVariable_subclass_clear` for consistency.

**Review by ezyang — APPROVED** [review]
whoops! Glad it was simple

**Review by kshitij12345 — APPROVED** [review]
Thank you very much for the quick fix @albanD

**Comment by kit1980** [comment]
2.5.1 is an emergency patch release to address known large regressions, moving this to 2.6.0

**Comment by ezyang** [comment]
@kit1980 if you're gonna move this fix to 2.6.0, you might as well say there's only one patch going in 2.5.1 lol. This one is clear, unambiguous and solves silent corruption (that happens rarely, but it is silent corruption!)

**Comment by kit1980** [comment]
@ezyang We were considering actually to do only one patch, but decided to include also couple of fixes for critical issues where external users loudly complained after 2.5.0.
We want to release 2.5.1 next week, and creating new release candidate will delay it and increase risk.
We might do 2.5.2 later.

**Comment by kit1980** [comment]
Also, is this a regression introduced in 2.5.0?
If not, then it's not for a patch release.

**Pull request closed without merging**